once/_iterator_wrappers.py

```diff
@@ -0,0 +1,186 @@
+import asyncio
+import collections.abc
+import threading
+
+# Before we begin, a note on the assert statements in this file:
+# Why are we using assert in here, you might ask, instead of implementing "proper" error handling?
+# In this case, it is actually not being done out of laziness! The assert statements here
+# represent our assumptions about the state at that point in time, and are always called with locks
+# held, so they **REALLY** should always hold. If the assumption behind one of these asserts fails,
+# the subsequent calls are going to fail anyways, so it's not like they are making the code
+# artificially brittle. However, they do make testing easer, because we can directly test our
+# assumption instead of having hard-to-trace errors, and also serve as very convenient
+# documentation of the assumptions.
+# We are always open to suggestions if there are other ways to achieve the same functionality in
+# python!
+
+
+class AsyncGeneratorWrapper:
+    """Wrapper around an async generator which only runs once.
+
+    Subsequent calls will return results from the first call, which is
+    evaluated lazily.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.AsyncGenerator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = asyncio.Lock()
+
+    async def _yield_results(self) -> collections.abc.AsyncGenerator:
```

Comment:
I'm okay with the `_iterator_wrappers` module name to communicate that it's private, but within that, IMO, `yield_results` should have no prefix because it's the public interface of this class. That way we can also have real private methods that even __init__.py shouldn't call.

Reply:
done